SAPHana/ra/SAPHanaController

```diff
@@ -1437,30 +1458,62 @@ function saphana_start() {
 }
 
 function saphana_stopSystem() {
-  local rc=0
-  check_sapstartsrv; rc=$?
-  if [ $rc -eq $OCF_SUCCESS ]; then
-     ## TODO: PRIO 1: Only stop System, if I am the last master!
-     ## TODO: PRIO 2: Do we need a "last-man-switch-of-the-light" detection?
-     output=$($SAPCONTROL -nr $InstanceNr -function StopSystem)
-     rc=$?
-     super_ocf_log info "ACT: SAPHANA STOP: Stopping System $SID-$InstanceName: $output"
-  fi
-  if [ $rc -eq 0 ]; then
-      HANA_ACTION_TIMEOUT=$(get_action_timeout)
-      output=$($SAPCONTROL -nr $InstanceNr -function WaitforStopped $HANA_ACTION_TIMEOUT 1)
-      if [ $? -eq 0 ]; then
-          super_ocf_log info "ACT: SAPHANA: System $SID-$InstanceName stopped: $output"
-          rc=$OCF_SUCCESS
-      else
-          super_ocf_log err "ACT: SAPHANA: System $SID-$InstanceName stop failed: $output, rc=$rc"
-          rc=$OCF_ERR_GENERIC
-      fi
-  else
-      super_ocf_log err "ACT: SAPHANA: System $SID-$InstanceName stop failed: $output, rc=$rc"
-      rc=$OCF_ERR_GENERIC
-  fi
-  return $rc
+    local rc=0 rcWfS=0
+    check_sapstartsrv; rc="$?"
+    if [ "$rc" -eq "$OCF_SUCCESS" ]; then
+        ## TODO: PRIO 1: Only stop System, if I am the last master!
+        ## TODO: PRIO 2: Do we need a "last-man-switch-off-the-light" detection?
+        output=$($SAPCONTROL -nr $InstanceNr -function StopSystem)
+        rc=$?
+         super_ocf_log info "ACT: SAP HANA STOP: Stopping System $SID-$InstanceName: $output"
+    fi
+    if [ "$rc" -eq 0 ]; then
+        HANA_ACTION_TIMEOUT=$(get_saphana_waitfor_timeout WaitforStopped)
+        output=$($SAPCONTROL -nr $InstanceNr -function WaitforStopped $HANA_ACTION_TIMEOUT 1); rcWfS="$?"
+        while true; do
+            # 
+            # If the master-nameserver dies and does not have a failover candidate but there are still running worker nodes
+            #    WaitforStopped terminates to early and does not wait till the worker nodes are down.
+            #    So we need to wait till the last "partial" node is down.
+            #
+            rcWfS=0
+            hanaANSWER=$(HANA_CALL --timeout $HANA_CALL_TIMEOUT --cmd "landscapeHostConfiguration.py --sapcontrol=1" 2>/dev/null); hanalrc="$?"
+            hAPart=$(echo "$hanaANSWER" | tr -d ' ' | \
+            awk -F= '
+               BEGIN { partials=0 }
+                $1 ~ "host/[a-zA-Z].*/hostActive" {hostActive = hostActive "-" $2; if ( $2 == "partial" ) { partials=1 } }
+                    END { printf "%s:%s\n", hostActive, partials;  } ' vName=$vName )
+            hanaHostActive="${hAPart%:*}"
+            partials="${hAPart#*:}"
+            logger -t "fhlpdebug:" "saphana_stop() hostActive=$hanaHostActive Partials=$partials"
+            if [ "$partials" -eq 0 ]; then
```

Comment:
if the HANA_CALL for landscapeHostConfiguration.py times out because of problems, the variable 'hanaANSWER' is empty. Because of the initialization within the 'awk' command starting in line 1482 we get a result ':0'. So in case that the HANA_CALL timed out, the variable 'partials' is 0, which leads to the result, that the stop was successful . Is this intended?
As we do not check the return value from the sapcontrol WaitforStopped in line 1472 (the variable is reset to 0 in the 'while true' loop starting in 1472) this can lead to a false positive.
So may be we need to check the return value of the HANA_CALL or we need to check, if 'hanaANSWER' is empty and reset the variable 'rcWfS' in regard to the result of this check or preserve the original value from the sapcontrol call.

Reply:
removed un-used return code: 87f1d6dcaeccf0c9af0c2381f92fd6e449cb4fef

Reply:
If WaitforStopped terminates we could have 3 stati:
a) The HANA is really down. In this case we also have no partial nodes so even if the loop would be by-passed by a 
    timed-out HANA call the result is correct. And with a stopped master-nameserver my guess is that the landscapeHostConfiguration.py does not time out. But lets see. rcWfS == 0 (loop rcWfL (we do not have that yet) is not neeed.
b) The HANA is not stopping correctly, then we (in deed) need the return code. rcWfS != 0, loop rcWfL (we do not have that yet) is not needed as we already know that we are failing.
c) The HANA is stopped partial rcWfS == 0, loop rcWfL (we do not have that yet) is needed to count the partial nodes (nodes still in shutdown).

==> So we need maybe 2 return codes (I tried to avoid that).
==> What should be the RA reaction if the second call times-out? Should we retry (in the loop)?